rust/main/chains/hyperlane-sovereign/src/provider/rest_client.rs

```diff
@@ -1,179 +1,1127 @@
+                    number,
+                })
+            } else {
+                Err(ChainCommunicationError::CustomError(String::from(
+                    "Bad response",
+                )))
+            }
+        } else {
+            Err(ChainCommunicationError::CustomError(String::from(
+                "Bad response",
+            )))
+        }
     }
 
-    pub async fn get_txn_by_hash(&self) -> ChainResult<TxnInfo> {
-        todo!()
+    // @Provider - test working
+    pub async fn get_txn_by_hash(&self, tx_hash: &H256) -> ChainResult<TxnInfo> {
+        info!(
+            "get_txn_by_hash(&self, tx_hash: &H256) tx_hash:{:?}",
+            tx_hash
+        );
+        #[derive(Clone, Debug, Deserialize)]
+        struct Data {
+            id: Option<String>,
+            _status: Option<String>,
+        }
+
+        // /sequencer/txs/{txHash}
+        let query = format!("/sequencer/txs/{tx_hash:?}");
+        // let query = format!("/sequencer/txs/{}", "0x2959329517b31126012eb858e33ae5b66ed466d67e4b6e722f1ef87b6f805b4a");
+
+        let response = self
+            .http_get(&query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Data> = serde_json::from_slice(&response)?;
+        println!("{response:?}",);
+
+        let res = TxnInfo {
+            hash: H256::from_str(
+                response
+                    .data
+                    .and_then(|d| d.id)
+                    .ok_or(ChainCommunicationError::CustomError(
+                        "Invalid response".to_string(),
+                    ))?
+                    .as_str(),
+            )?,
+            gas_limit: U256::default(),
+            max_priority_fee_per_gas: Some(U256::default()),
+            max_fee_per_gas: Some(U256::default()),
+            gas_price: Some(U256::default()),
+            nonce: u64::default(),
+            sender: H256::default(),
+            recipient: Some(H256::default()),
+            receipt: Some(TxnReceiptInfo {
+                gas_used: U256::default(),
+                cumulative_gas_used: U256::default(),
+                effective_gas_price: Some(U256::default()),
+            }),
+        };
+        Ok(res)
     }
 
-    pub async fn is_contract(&self) -> ChainResult<bool> {
-        todo!()
+    pub async fn get_batch(&self, batch: u64) -> ChainResult<Batch> {
+        info!("get_batch_tx_event(&self, batch: u64) batch:{:?}", batch);
+        let query = format!("/ledger/batches/{batch}?children=1");
+
+        let response = self
+            .http_get(&query)
+            .await
+            .map_err(|e| ChainCommunicationError::CustomError(format!("HTTP Get Error: {e}")))?;
+        let response: Schema<Batch> = serde_json::from_slice(&response)?;
+        let data = response.data.ok_or(ChainCommunicationError::CustomError(
+            "Invalid response".to_string(),
+        ))?;
+        Ok(data)
```

Comment:
Would it be simplier?

```suggestion
       response
           .data
           .ok_or_else(|| ChainCommunicationError::CustomError("Invalid response: missing batch data".to_string()))
```